common/dictionary-db/dictionary-db.ts

```diff
@@ -0,0 +1,1416 @@
+                                for (const lemma of record.lemmas) modifiedTokens.add(lemma);
+                                record.cardIds = remainingCardIds;
+                            }
+                        }),
+                    this.db.ankiCards
+                        .where('[cardId+track+profile]')
+                        .anyOf(orphanedCardIds.map((cardId) => [cardId, track, profile]))
+                        .delete(),
+                ]);
+            });
+        });
+    }
+
+    private async _orphanAllCardIds(profile: string, tracks: number[]): Promise<Map<number, number[]>> {
+        if (!tracks.length) return new Map();
+        const orphanedTrackCardIds = new Map<number, number[]>(tracks.map((track) => [track, []]));
+        return this._getAnkiCards(profile).then((ankiCards) => {
+            for (const ankiCard of ankiCards) {
+                const arr = orphanedTrackCardIds.get(ankiCard.track);
+                if (!arr) continue;
+                arr.push(ankiCard.cardId);
+            }
+            return orphanedTrackCardIds;
+        });
+    }
+
+    private async _ensureBuildId(
+        key: DictionaryMetaKey,
+        buildId: string,
+        options?: { buildTs: number }
+    ): Promise<boolean> {
+        return this.db.transaction('rw', this.db.meta, async () => {
+            const trackMeta = await this.db.meta.where('[profile+track]').equals(key).first();
+            if (!options?.buildTs) return trackMeta?.buildId === buildId; // For health check
+            const { buildTs } = options;
+            const initialExpiration = buildTs + BUILD_MIN_EXPIRATION_MS; // Will be updated continuously as work is done
+
+            // First time build for this track
+            if (!trackMeta) {
+                await this.db.meta.add({
+                    profile: key[0],
+                    track: key[1],
+                    lastBuildStartedAt: buildTs,
+                    lastBuildExpiresAt: initialExpiration,
+                    buildId,
+                    settings: null,
+                });
+                return true;
+            }
+
+            // Build already in progress, ensure it's not stale
+            if (trackMeta.buildId && trackMeta.buildId !== buildId) {
+                if (buildTs < trackMeta.lastBuildExpiresAt) return false;
+                console.warn(
+                    `Stale buildId ${trackMeta.buildId} which expired at ${new Date(trackMeta.lastBuildExpiresAt).toISOString()} detected for track ${key[1] + 1}, ignoring.`
+                );
+            }
+
+            await this.db.meta.update(key, {
+                lastBuildStartedAt: buildTs,
+                lastBuildExpiresAt: initialExpiration,
+                buildId,
+            });
+            return true;
+        });
+    }
+
+    private async _buildIdHealthCheck(buildId: string, trackBuildIdsToClear: DictionaryMetaKey[]): Promise<void> {
```

Comment:
Suggestion to use a different name than `trackBuildIdsToClear`. I see that build IDs do get cleared from the database in some places, but it makes less sense in the context of this method.